compiler/src/main/java/com/dylibso/chicory/compiler/internal/Emitters.java

```diff
@@ -658,4 +668,78 @@ public static BytecodeEmitter intrinsify(CompilerOpCode opcode, Class<?> staticH
                         + " does not provide an implementation of opcode "
                         + opcode.name());
     }
+
+    public static void THROW(Context ctx, CompilerInstruction ins, InstructionAdapter asm) {
+        int tagNumber = (int) ins.operand(0);
+        var type = ctx.tagFunctionType(tagNumber);
+
+        // emmit:
+        // call createWasmException(long[] args, int tagNumber, Instance instance)
+        emitBoxValuesOnStack(ctx, asm, type.params());
+        asm.iconst(tagNumber);
+        asm.load(ctx.instanceSlot(), OBJECT_TYPE);
+        emitInvokeStatic(asm, ShadedRefs.CREATE_WASM_EXCEPTION);
+        asm.athrow();
+    }
+
+    public static void THROW_REF(Context ctx, CompilerInstruction ins, InstructionAdapter asm) {
+        // The exception reference is already on the stack as an integer
+        // Get the instance and retrieve the exception
+        asm.load(ctx.instanceSlot(), OBJECT_TYPE);
+        asm.swap(); // Swap instance and exception reference
+        emitInvokeVirtual(asm, ShadedRefs.INSTANCE_GET_EXCEPTION);
+        asm.athrow();
+    }
+
+    public static void CATCH_START(Context ctx, CompilerInstruction ins, InstructionAdapter asm) {
+        asm.store(ctx.tempSlot(), OBJECT_TYPE);
+    }
+
+    public static void CATCH_END(Context ctx, CompilerInstruction ins, InstructionAdapter asm) {
+        asm.load(ctx.tempSlot(), OBJECT_TYPE);
+        asm.athrow();
+    }
+
+    public static void CATCH_UNBOX_PARAMS(
+            Context ctx, CompilerInstruction ins, InstructionAdapter asm) {
+        var tag = (int) ins.operand(0);
+        // Get the tag type to know what
+        // parameter types to unbox
+        var tagFuncType = ctx.tagFunctionType(tag);
+        if (!tagFuncType.params().isEmpty()) {
+            // unbox the exception args
+            asm.load(ctx.tempSlot(), OBJECT_TYPE);
+            asm.invokevirtual(
+                    getInternalName(WasmException.class),
+                    "args",
+                    getMethodDescriptor(getType(long[].class)),
+                    false);
+
+            // Store the array in a local variable
+            // Unbox each argument from the
+            // long[] array and push onto stack
+            emitUnboxResult(asm, tagFuncType.params(), ctx.tempSlot() + 1);
```

Comment:
not sure 100% why/how ctx.tempSlot() **+1** works? why can't we reuse tempSlot here? 🤔